search/typesense.py

```diff
@@ -0,0 +1,533 @@
+
+To switch to this engine, you should run a Typesense instance and define the
+following setting:
+
+    SEARCH_ENGINE = "search.typesense.TypesenseEngine"
+
+You will then need to create the new indices by running:
+
+    ./manage.py lms shell -c "import search.typesense; search.typesense.create_indexes()"
+
+^ This manual index creation is a temporary approach, while we're evaluating
+Typesense. A proper long-term approach is needed for all our search engines, as
+discussed at https://github.com/openedx/edx-platform/issues/36868 .
+
+For more information about the Typesense API in Python, check
+https://github.com/typesense/typesense-python
+"""
+import logging
+import typing as t
+
+from django.conf import settings
+from django.utils import timezone
+import typesense
+from typesense.collection import Collection
+from typesense.exceptions import ObjectNotFound, RequestMalformed, ServerError
+from typesense.types.collection import RegularCollectionFieldSchema
+from typesense.types.document import Hit, SearchResponse
+
+from search.search_engine_base import SearchEngine
+from search.utils import (
+    convert_doc_datatypes,
+    IS_NULL_SUFFIX,
+    restore_doc_datatypes,
+    UTC_OFFSET_SUFFIX,
+    ValueRange,
+)
+
+# ====== Indices ======
+# Due to the messy architecture of edx-search, we have to encode information about specific indexes in this engine.
+# So some information about the indexes is here (in ES, Meilisearch, and Typesense engines separately) and other info
+# is in edx-platform, e.g. edx-platform/cms/djangoapps/contentstore/courseware_index.py
+#
+# The two indexes we're most concerned about are:
+#  - Course Info index: stores a list of courses that users can browse and then enroll into
+#  - Course Content index: stores the course content (XBlocks) for enrolled users to search within one course
+# We need to know which fields store timestamps in order to force them to be stored as 64 bit integers instead of 32
+COURSE_INFO_INDEX = getattr(settings, "COURSEWARE_INFO_INDEX_NAME", "course_info")
+COURSE_CONTENT_INDEX = getattr(settings, "COURSEWARE_CONTENT_INDEX_NAME", "courseware_content")
+INDEX_CONFIGURATION = {
+    COURSE_INFO_INDEX: {
+        # Most fields will be auto-created but some (datetimes, facet fields) need to be specified up front:
+        "field_overrides": [
+            {"name": "start", "type": "datetime", "optional": True},
+            {"name": "end", "type": "datetime", "optional": True},
+            {"name": "enrollment_start", "type": "datetime", "optional": True},
+            {"name": "enrollment_end", "type": "datetime", "optional": True},
+            {"name": "org", "type": "string", "facet": True},
+            {"name": "modes", "type": "string[]", "facet": True},
+            {"name": "language", "type": "string", "facet": True},
+        ],
+        # Which fields to use for text matches. Required by Typesense.
+        "query_fields": ["course", "org", "number", "content"],
```

Comment:
@bradenmacdonald sorry, can we add stemming for the course and content fields here too?

Reply:
Ah my bad, yeah the `course` field is the course ID, while the `content` field is an object with fields that includes the overview and display name. Maybe we can add stemming for the `content` field, similarly to the content field in the other index?

Reply:
Sure, done: 4efeab7c6266280ccce3c011ad5150989552c2cb

Reply:
Hmm, it's a little hard to know what is covered by "stemming" and what is covered by "typo tolerance", and what is excluded due to short word length. All three are factors in any given thing. But I do think this should be more similar to Meilisearch now.